### ISSUE TITLE
Timing conflict when initializing Ubuntu 16.04 VPS on Vultr

### ISSUE THREAD

**Issue by fake-name** [saltstack/salt#37062]
### Description of Issue/Question

Basically, the vultr VPS deployment procedure involves some `apt-get` management. Basically, the deploy process works like this.
- VPS Is created. 
- Initial setup is done. VPS Is not accessible.
- VPS is treated as "created". API call returns "ok".
- VPS Is now accessible over SSH.
- Deploy script from Vultr continues to run in the background, executing something like `apt-get update && apt-get upgrade`.
- Several minutes later, the update finishes in the background.

Critically, the VPS becomes SSH accessible several minutes _before_ the setup is complete. As a result, the salt deploy script starts executing _while_ the deployment is still ongoing. As a result, calls to `apt` from the salt deploy script can collide with the vultr deploy script, since `apt` requires an exclusive lock.

The console output looks like: 

```
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Permission denied, please try again.
root@45.32.101.19's password:
Tue Oct 18 03:18:11 UTC 2016
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connection to 45.32.101.19 closed.
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connection to 45.32.101.19 closed.
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connected to 45.32.101.19.
sftp> put  /tmp/tmprze4Tc /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion.pem
Uploading /tmp/tmprze4Tc to /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion.pem
/tmp/tmprze4Tc                                                                                                                        100% 1678     1.6KB/s   00:00
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connection to 45.32.101.19 closed.
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connected to 45.32.101.19.
sftp> put  /tmp/tmp6l8JnI /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion.pub
Uploading /tmp/tmp6l8JnI to /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion.pub
/tmp/tmp6l8JnI                                                                                                                        100%  450     0.4KB/s   00:01
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connected to 45.32.101.19.
sftp> put  /tmp/tmptbbld3 /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion
Uploading /tmp/tmptbbld3 to /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/minion
/tmp/tmptbbld3                                                                                                                        100%   75     0.1KB/s   00:00
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connected to 45.32.101.19.
sftp> put  /tmp/tmpTD54rJ /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/deploy.sh
Uploading /tmp/tmpTD54rJ to /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/deploy.sh
/tmp/tmpTD54rJ                                                                                                                        100%  242KB 241.9KB/s   00:01
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
Connection to 45.32.101.19 closed.
Warning: Permanently added '45.32.101.19' (ECDSA) to the list of known hosts.
root@45.32.101.19's password:
 *  INFO: Running version: 2016.08.16
 *  INFO: Executed by:
 *  INFO: Command line: '/tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/deploy.sh -c /tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb'

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   4.4.0-36-generic
 *  INFO:   Distribution: Ubuntu 16.04

 *  INFO: Installing minion
 *  INFO: Found function install_ubuntu_stable_deps
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function install_ubuntu_stable
 *  INFO: Found function install_ubuntu_stable_post
 *  INFO: Found function install_ubuntu_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Found function install_ubuntu_check_services
 *  INFO: Running install_ubuntu_stable_deps()
Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [94.5 kB]
Hit:2 http://archive.ubuntu.com/ubuntu xenial InRelease
Hit:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease
Hit:4 http://archive.ubuntu.com/ubuntu xenial-backports InRelease
Fetched 94.5 kB in 1s (71.6 kB/s)
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [94.5 kB]
Hit:2 http://archive.ubuntu.com/ubuntu xenial InRelease
Hit:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease
Connection to 45.32.101.19 closed.
Hit:4 http://archive.ubuntu.com/ubuntu xenial-backports InRelease
Fetched 94.5 kB in 1s (74.2 kB/s)
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
Traceback (most recent call last):
  File "salt-test.py", line 17, in <module>
    testf()
  File "salt-test.py", line 12, in testf
    size               = u'29'
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 408, in create
    mapper.create(vm_))
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1281, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/vultrpy.py", line 380, in create
    ret = __utils__['cloud.bootstrap'](vm_, __opts__)
  File "/usr/lib/python2.7/dist-packages/salt/utils/cloud.py", line 527, in bootstrap
    deployed = deploy_script(**deploy_kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/utils/cloud.py", line 1516, in deploy_script
    if root_cmd(deploy_command, tty, sudo, **ssh_kwargs) != 0:
  File "/usr/lib/python2.7/dist-packages/salt/utils/cloud.py", line 2167, in root_cmd
    retcode = _exec_ssh_cmd(cmd, allow_failure=allow_failure, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/utils/cloud.py", line 1784, in _exec_ssh_cmd
    cmd, proc.exitstatus
salt.exceptions.SaltCloudSystemExit: Command 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@45.32.101.19 '/tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb/deploy.sh -c '"'"'/tmp/.saltcloud-17bd50f5-7da6-43b4-a34d-5381c0be42eb'"'"''' failed. Exit code: 1
1 durr@rwpscrape:/media/Storage/Scripts/AutoTriever/marshaller⟫

```
### Setup

This shouldn't be too relevant,

In any event, this is using the exact test setup from https://github.com/saltstack/salt/issues/37040, after updating `vultrpy.py` to `HEAD`.
### Steps to Reproduce Issue

Execute 

```
import salt.cloud

def testf():

    cc = salt.cloud.CloudClient('/etc/salt/cloud')
    cc.create(names        = ['test-1'],
        provider           ='vultr',
        private_networking = False,
        image              = 'Ubuntu 16.04 x64',
        location           = 40,
        size               = u'29'
        )

if __name__ == '__main__':

    testf()

```

Against a vultr account.
#### Expected behaviour:
- VPS is created.
#### Actual behaviour:
- VPS creation fails due to apt lock conflict.
### Versions Report

See https://github.com/saltstack/salt/issues/37040

---

Realistically, simply retrying on apt-fail for a period of time, or allowing apt-get update/upgrade to block until the background process has finished should fix the issue.

**Comment by fake-name** [comment]
Note that this is a random failure thing. Out of the 5 attempted deploys I've done so far, I've had it succeed once. I'm currently poking about with modifying the `deploy.sh` script.

**Comment by fake-name** [comment]
Sticking 

```
    while fuser /var/lib/dpkg/lock >/dev/null 2>&1 ; do
        sleep 0.5
    done
```

in `bootstrap-salt.sh` seems to fix the issue. 

In this case, I amended 

```
install_ubuntu_deps() {
    
    apt-get update
```

to

```
install_ubuntu_deps() {

    while fuser /var/lib/dpkg/lock >/dev/null 2>&1 ; do
        sleep 0.5
    done

    apt-get update
```

which seems to resolve the issue.

While this seems harmless, I'm not sure if it something that would be added in general. I'd be happy to pull request this and https://github.com/saltstack/salt/issues/37063. Should I?

**Comment by Ch3LL** [comment]
@fake-name just to make sure do you happen to be using any `startup_states`? Just want to make sure you are not running into #36599

Also if you run salt-cloud with `script_args: -X` does it resolve this issue? Thanks

**Comment by fake-name** [comment]
I don't even know what a `startup_state` is, so I assume not.  Basically, the above demo script is run on a clean install of `salt-cloud`, with the only change to the configuration being the set-up of my VPS API credentials.

What does "script_args: -X" do? Note that I'm interacting **entirely** via the python API, so I have no yaml anything. I think I can pass that through to the  `CloudClient.create()` call (I've been passing `script_args="-D"` in for better debug output), but I'm not sure. if that's the same place.

**Comment by fake-name** [comment]
Found the help, `-X` appears to not start the salt daemon? I need it running (I do stuff through it).

**Comment by Ch3LL** [comment]
apologies I should have been more clear I would have wanted you to try `-X` if you were using `startup_states` as a temporary workaround. That is a separate issue where the states would be run and start attempting to update packages when the minion was started.

So since you are definitely not using startup_states that is not related. I just wanted to rule that out quickly. Thanks for clarifying. 

Let me ping someone real quick to find out if we want to handle this in the code or the bootstrap script like you have. Thanks

**Comment by gtmanfred** [comment]
Does vultr have a tag that the put on servers to let you know when their automation is complete?

With rackspace, their managed automation adds the following tag to the metadata of the server so that we know that we can run our automation after it.  

https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/nova.py#L800
and
https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/nova.py#L810

We will only bootstrap the server if the rackconnect and managed automation statuses are complete.

If vultr has something like that, we would rather do add an option like the managed or rackconnect options in the nova and openstack drivers, instead of adding something extra like what was mentioned above into salt-bootstrap.

**Comment by fake-name** [comment]
Not that I can see, unfortunately. A lot of their API stuff seems kinda hacked-together. For instance, they generate an e-mail for *every* instance created, and there isn't a way to turn the notification off *at all*. Their support literally suggested I just use e-mail filters.

Their interface also notifies you "updates may occur in the background for the first few minutes" when you create a new server, so I think they kind of just tell you, and hope you work around it.

**Comment by gtmanfred** [comment]
Ok, would you open a support ticket with them just to make sure that there isn't anything else that they can do about providing information to their users about when their first boot automation is finished.

If they won't do that, then we should open an issue on https://github.com/saltstack/salt-bootstrap and discuss the best way to solve it there.

**Comment by fake-name** [comment]
Will do. It'll have to wait until tonight when I can get into my account (I'm at work now, and I don't have access to my account credentials on this computer).

**Comment by gtmanfred** [comment]
Awesome, thanks!

**Comment by fake-name** [comment]
Opened an support ticket. We'll see what they say.

**Comment by fake-name** [comment]
So they got back to me:

> Hello,
> 
> Thank you for contacting us!
> 
> I am fairly sure that the "status" entry from the below linked API call JSON response should provide the flag that you require. Status should return "active" once all provisioning is complete. If I'm not mistaken, the status should be "pending" or "installing" if provisioning is still taking place.
> 
> https://www.vultr.com/api/#server_server_list
> 
> Please feel free to let us know if you find otherwise and we'll be sure to have the development team have a look.

It sounds like I can poke another part of the API, and get the state. I'm going to see about having a look in the `vultrpy.py` module. Hopefully this weekend.

**Comment by fake-name** [comment]
Another potential issue that should possibly be looked into (another bug) is the fact that you can create multiple instances of a VM with the same name on vultr. It makes the client very confused, because which instance you get is a function of the response layout, and what python decides to do with the dictionary generated from it. The ordering is non-deterministic.

**Comment by fake-name** [comment]
Ok, Now I'm extra confused. If I ssh into a new server immediately when it's created, it's idle.

If I let the bootstrap process run to the point that it ssh-es into the new server, I see apt running. I'm not sure /who/ is executing apt. I wonder if the SSH connection is starting apt, and closing the ssh session before the command completes, which then continues to run in the background.

This is all more annoying to troubleshoot, because I have to start up a new vm to test everything. Arrrgh.

**Comment by fake-name** [comment]
Ok, I'm not sure what's causing this issue. If I manually create a vultr VPS, and ssh into it as soon as it's created, I don't see any instances of apt running. On the other hand, I can pretty easily replicate the apt conflict described in the first post of this issue pretty easily, 

I modified `vultrpy` to wait for the additional flags the vultr help people got back to me about, and it doesn't seem to rectify the issues. At this point, I think the issue is actually a timing issue in the deploy system, probably some interaction between a vagary of the default ubuntu image from vultr and the salt-deploy script.

On the other hand, I've had 100% success with my deploys with the modifications that simply wait for the dpkg lock, so maybe it'd be worth implementing anyways. I can't see it hurting, and it makes the installation process more robust. Thoughts?

**Comment by Ch3LL** [comment]
Perfect thanks for the fix. I'll go ahead and close this issue. I agree this is definitely a different and appropriate way to approach the issue especially if there are concerns about the returns from teh vultr api. Thanks for the fix.

**Comment by fake-name** [comment]
To be clear, the #37170 alone **does not** fix this issue. I'm not sure what the current tags mean (does "blocked" mean it's a genuine issue, but not solvable atm?).

The additional extensions to `bootstrap-salt.sh` [above](https://github.com/saltstack/salt/issues/37062#issuecomment-254403310) do fix the issue, but in a rather brute-force way, which is what I was interested in comments on. I'll open a bug report on salt-bootstrap if that would be appropriate, and the changes aren't too left-field.

**Comment by gtmanfred** [comment]
Yeah, we need to leave this open.

What I would say, you should let vultry know what changes you made, and that they are still running their deployment scripts after the server goes into active state.  The really correct way to fix it is what you did, but if they are not giving us any information on what to do, it may require adding another step into the vultr create function to login and run that check that you are running above.

I think that that would probably be a better solution than including it in salt-bootstrap.sh

**Comment by fake-name** [comment]
The issue I have here is that I spent a little while trying to see what vultr is doing, and I can't actually see what it is. If I ssh into a new VM on them as soon as it's available and run `ps`, I don't *see* anything running. I'm not sure if the issue is actually something Vultr is doing, or some bizarre interaction between something in vultr's ubuntu image and the deploy script.

**Comment by gtmanfred** [comment]
I am going to try and take a look into it today or monday when I have time.

**Comment by fake-name** [comment]
Ok, even with my "check-if-dpkg-lock" stuff, I'm *still* getting failure-to-create issues with vultr:

```
Warning: Permanently added '45.32.160.106' (ECDSA) to the list of known hosts.
root@45.32.160.106's password:
 *  INFO: Running version: 2016.08.16
 *  INFO: Executed by:
 *  INFO: Command line: '/tmp/.saltcloud-532ba37b-b4a0-4d6d-ba9e-bd92310b9df7/deploy.sh -c /tmp/.saltcloud-532ba37b-b4a0-4d6d-ba9e-bd92310b9df7 -D'

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   4.4.0-45-generic
 *  INFO:   Distribution: Ubuntu 16.04

 * DEBUG: Binaries will be searched using the following $PATH: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
 *  INFO: Installing minion
 * DEBUG: install_ubuntu_16_stable_deps not found....
 * DEBUG: install_ubuntu_16_04_stable_deps not found....
 * DEBUG: install_ubuntu_16_deps not found....
 * DEBUG: install_ubuntu_16_04_deps not found....
 *  INFO: Found function install_ubuntu_stable_deps
 * DEBUG: DEPS_INSTALL_FUNC=install_ubuntu_stable_deps
 * DEBUG: config_ubuntu_16_stable_salt not found....
 * DEBUG: config_ubuntu_16_04_stable_salt not found....
 * DEBUG: config_ubuntu_16_salt not found....
 * DEBUG: config_ubuntu_16_04_salt not found....
 * DEBUG: config_ubuntu_stable_salt not found....
 * DEBUG: config_ubuntu_salt not found....
 *  INFO: Found function config_salt
 * DEBUG: CONFIG_SALT_FUNC=config_salt
 * DEBUG: preseed_ubuntu_16_stable_master not found....
 * DEBUG: preseed_ubuntu_16_04_stable_master not found....
 * DEBUG: preseed_ubuntu_16_master not found....
 * DEBUG: preseed_ubuntu_16_04_master not found....
 * DEBUG: preseed_ubuntu_stable_master not found....
 * DEBUG: preseed_ubuntu_master not found....
 *  INFO: Found function preseed_master
 * DEBUG: PRESEED_MASTER_FUNC=preseed_master
 * DEBUG: install_ubuntu_16_stable not found....
 * DEBUG: install_ubuntu_16_04_stable not found....
 *  INFO: Found function install_ubuntu_stable
 * DEBUG: INSTALL_FUNC=install_ubuntu_stable
 * DEBUG: install_ubuntu_16_stable_post not found....
 * DEBUG: install_ubuntu_16_04_stable_post not found....
 * DEBUG: install_ubuntu_16_post not found....
 * DEBUG: install_ubuntu_16_04_post not found....
 *  INFO: Found function install_ubuntu_stable_post
 * DEBUG: POST_INSTALL_FUNC=install_ubuntu_stable_post
 * DEBUG: install_ubuntu_16_stable_restart_daemons not found....
 * DEBUG: install_ubuntu_16_04_stable_restart_daemons not found....
 * DEBUG: install_ubuntu_16_restart_daemons not found....
 * DEBUG: install_ubuntu_16_04_restart_daemons not found....
 * DEBUG: install_ubuntu_stable_restart_daemons not found....
 *  INFO: Found function install_ubuntu_restart_daemons
 * DEBUG: STARTDAEMONS_INSTALL_FUNC=install_ubuntu_restart_daemons
 * DEBUG: daemons_running_ubuntu_16_stable not found....
 * DEBUG: daemons_running_ubuntu_16_04_stable not found....
 * DEBUG: daemons_running_ubuntu_16 not found....
 * DEBUG: daemons_running_ubuntu_16_04 not found....
 * DEBUG: daemons_running_ubuntu_stable not found....
 * DEBUG: daemons_running_ubuntu not found....
 *  INFO: Found function daemons_running
 * DEBUG: DAEMONS_RUNNING_FUNC=daemons_running
 * DEBUG: install_ubuntu_16_stable_check_services not found....
 * DEBUG: install_ubuntu_16_04_stable_check_services not found....
 * DEBUG: install_ubuntu_16_check_services not found....
 * DEBUG: install_ubuntu_16_04_check_services not found....
 * DEBUG: install_ubuntu_stable_check_services not found....
 *  INFO: Found function install_ubuntu_check_services
 * DEBUG: CHECK_SERVICES_FUNC=install_ubuntu_check_services
 *  INFO: Running install_ubuntu_stable_deps()
Waiting on another DPKG process to complete!
Waiting on another DPKG process to complete!
Waiting on another DPKG process to complete!
Waiting on another DPKG process to complete!
Waiting on another DPKG process to complete!
Waiting on another DPKG process to complete!
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
Connection to 45.32.160.106 closed.
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
 * DEBUG: Removing the logging pipe /tmp/bootstrap-salt.logpipe
 * DEBUG: Killing logging pipe tee's with pid(s): 2311
```

I have no idea why.

**Comment by gtmanfred** [comment]
I am going to try and take a look this week at this issue.

**Comment by fake-name** [comment]
To be clear, I'm not trying to be pushy or anything, it's just if I don't note things down when I see them, I'll forget whatever it was by the time I actually can look at them. I'm basically that dog from the "Up" movie .... Squirrel!

**Comment by gtmanfred** [comment]
:) no problem! I really do appreciate all the leg work you have done on this so far!

**Comment by fake-name** [comment]
FWIW, I just encountered this issue still, after updating to the latest salt:

```
 *  INFO: Installing minion
 *  INFO: Found function install_ubuntu_stable_deps
 *  INFO: Found function config_salt
 *  INFO: Found function install_ubuntu_stable
 *  INFO: Found function install_ubuntu_stable_post
 *  INFO: Found function install_ubuntu_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Found function install_ubuntu_check_services
 *  INFO: Running install_ubuntu_stable_deps()
Hit:1 http://archive.ubuntu.com/ubuntu xenial InRelease
Hit:2 http://security.ubuntu.com/ubuntu xenial-security InRelease
Hit:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease
Connection to 45.32.117.82 closed.
Hit:4 http://archive.ubuntu.com/ubuntu xenial-backports InRelease
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
Traceback (most recent call last):
  File "salt_runner_test.py", line 19, in <module>
    testf()
  File "salt_runner_test.py", line 12, in testf
    size               = u'29'
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 409, in create
    mapper.create(vm_))
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1285, in create
    output = self.clouds[func](vm_)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/clouds/vultrpy.py", line 413, in create
    ret = __utils__['cloud.bootstrap'](vm_, __opts__)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/cloud.py", line 532, in bootstrap
    deployed = deploy_script(**deploy_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/cloud.py", line 1524, in deploy_script
    if root_cmd(deploy_command, tty, sudo, **ssh_kwargs) != 0:
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/cloud.py", line 2221, in root_cmd
    retcode = _exec_ssh_cmd(cmd, allow_failure=allow_failure, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/cloud.py", line 1802, in _exec_ssh_cmd
    cmd, proc.exitstatus
salt.exceptions.SaltCloudSystemExit: Command 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@45.32.117.82 '/tmp/.saltcloud-0ed1bf17-7e54-4cf3-b3f6-09fd5cf18f2c/deploy.sh -c '"'"'/tmp/.saltcloud-0ed1bf17-7e54-4cf3-b3f6-09fd5cf18f2c'"'"''' failed. Exit code: 1
1 client@rabbit:~/AutoTriever/marshaller⟫ sudo python salt_runner.py destroy-all
Setting up loggers....
done
Main.VpsHerder.MainThread - INFO - Active nodes: ['test-1']
```

The hacked up version of `bootstrap-salt.sh` seems to have solved the problem to a reasonable extent, though.

**Comment by gtmanfred** [comment]
Could you check if this PR fixes the issue?

https://github.com/saltstack/salt/pull/40548

Thanks,
Daniel

**Comment by fake-name** [comment]
I don't see how that PR would have any effect at all on this issue.

That PR fixes the fact that Vultr returns `0.0.0.0` for the IP before the server IP is known. This issue is due to the fact that Vultr has ongoing `apt-get` activity that continues even after ssh access is available, which causes `bootstrap-salt.sh` to fail.

**Comment by gtmanfred** [comment]
Ok, was just checking, since i saw that it was another possible race condition in vultr.

Thanks,
Daniel

**Comment by nicosmaris** [comment]
As a workaround, I use the following script to kill the apt command. This is because I always add more sources and then run 'apt-get update', so letting the initial 'apt-get update' command finish is useless.
```
  killall -we 'apt-get'
  echo "start searching for 'apt-get -qq -y update' initiated by vultr..."
  ps aux | grep '[a]pt'
  echo "stop searching for 'apt-get -qq -y update' initiated by vultr..."
  rm /var/lib/apt/lists/lock
  rm /var/cache/apt/archives/lock
  rm /var/lib/dpkg/lock
  dpkg --configure -a
```

**Comment by KumarRajnish** [comment]
All,

Due to a misconfiguration on our Office365 tenant, all email addressed to your Ruckus Wireless email address (you@ruckuswireless.com) were delivered to a catch all mailbox.  This incident occurred between 11 PM to 8 AM (PST).  IT is delivering these emails with a “Reply-all” rule to all individuals on the email chain.  If you feel you are missing email attachments, please send email to helpdesk@ruckuswireless.com.

We apologize for any inconvenience this may have caused.

Regards,

IT

**Comment by fake-name** [comment]
I'm still [running my workaround for this](https://github.com/fake-name/AutoTriever/blob/master/marshaller/salt_runner.py#L217-L229).

**Comment by fake-name** [comment]
I'm /still/ running my patch.

Is salt ded?

**Comment by whytewolf** [comment]
@fake-name You mentioned you can created a PR. That would be greatly appreciated. however because of the age of this ticket and some of changes with the way we track things including how our automation works. we have to close this ticket.

If you have the time to put in the PR please open a new ticket with the new template. and put together the PR with a changelog pointing at that issue.

**Comment by fake-name** [comment]
It's always nice to see the [CADT](https://www.jwz.org/doc/cadt.html) development model in action.

I need to move away from saltstack.
